### ISSUE TITLE
getIsAuthorized() is never true.

### ISSUE THREAD

**Issue by jbloem9** [damienbod/AspNet5IdentityServerAngularImplicitFlow#86]
When logging in identity server is called and credentials are entered ,after returning from ids user is never set as authorized(.getIsAuthorized() always returns false).
.getToken() is always blank
[user.service.txt](https://github.com/damienbod/AspNet5IdentityServerAngularImplicitFlow/files/1189483/user.service.txt)
setupModule:
constructor(public oidcSecurityService: OidcSecurityService) {
        let openIDImplicitFlowConfiguration = new OpenIDImplicitFlowConfiguration();
        openIDImplicitFlowConfiguration.stsServer = environment.stsServer;
        openIDImplicitFlowConfiguration.redirect_url = environment.redirect_url;
        openIDImplicitFlowConfiguration.client_id = 'AuctionX.Web.Ui.Client.Ng';
        openIDImplicitFlowConfiguration.response_type = 'id_token token';
        openIDImplicitFlowConfiguration.scope = 'openid email profile role api.write api.read';
        openIDImplicitFlowConfiguration.post_logout_redirect_uri = environment.post_logout_redirect_uri;
        openIDImplicitFlowConfiguration.start_checksession = false;
        openIDImplicitFlowConfiguration.silent_renew = true;
        openIDImplicitFlowConfiguration.startup_route = '/home';
        openIDImplicitFlowConfiguration.forbidden_route = '/forbidden';
        openIDImplicitFlowConfiguration.unauthorized_route = '/unauthorized';
        openIDImplicitFlowConfiguration.log_console_warning_active = true;
        openIDImplicitFlowConfiguration.log_console_debug_active = false;
        openIDImplicitFlowConfiguration.max_id_token_iat_offset_allowed_in_seconds = 10;
        openIDImplicitFlowConfiguration.override_well_known_configuration = false;
        openIDImplicitFlowConfiguration.override_well_known_configuration_url = environment.well_known_config_url;
        openIDImplicitFlowConfiguration.storage = localStorage;

        this.oidcSecurityService.setupModule(openIDImplicitFlowConfiguration);
    }

**Comment by MovGP0** [comment]
what URL/Headers do you get back from the server?

**Comment by jbloem9** [comment]
When the application loads we have the following in the console, see 
![loginstart](https://user-images.githubusercontent.com/7578265/28948280-baba2974-7908-11e7-9252-3bc443307dc5.PNG)

Then I was getting the same issue as angular-auth-oidc-client issue number #50.  Implemented the changes suggested viz implemented the subscription to onModuleSetup before calling authorize.  However now the issues is we never get a result back from onModuleSetup and as such the login page is never brought up.  I have reattached my changed code. 
[UserService.txt](https://github.com/damienbod/AspNet5IdentityServerAngularImplicitFlow/files/1198897/UserService.txt)

**Comment by MovGP0** [comment]
the TLS (aka SSL) certificate is not valid

**Comment by MovGP0** [comment]
use this to create your test keys; you only need to replace the „mydomain“, folders and the passwords with something of your own: 
```powershell
function New-TestCertificates
{
	param()
	process
	{
		[string]$basePath = (Join-Path $PSScriptRoot '..' -Resolve);

		Write-Information 'Creating Root Certificate';

		$rootCert = New-SelfSignedCertificate -CertStoreLocation cert:\localmachine\my -DnsName "MyDomainCertificate Authority" -KeyUsage CertSign -KeyUsageProperty All;

		Write-Information 'Exporting Root Certificate';

		[System.Security.SecureString]$rootcertPassword = ConvertTo-SecureString -String "PUT SOME REALLY SECURE PASSWORD HERE" -Force -AsPlainText;
		[String]$rootCertPath = Join-Path -Path 'cert:\localMachine\my\' -ChildPath 
"MyDomain Certificate Authority" -KeyUsage CertSign -KeyUsageProperty All;

		Write-Information 'Exporting Root Certificate';

		[System.Security.SecureString]$rootcertPassword = ConvertTo-SecureString -String "PUT SOME REALLY SECURE PASSWORD HERE" -Force -AsPlainText;
		[String]$rootCertPath = Join-Path -Path 'cert:\localMachine\my\' -ChildPath "$($rootcert.Thumbprint)";
		Export-PfxCertificate -Cert $rootCertPath -FilePath (Join-Path $basePath 'root.pfx') -Password $rootcertPassword; # private key
		Export-Certificate    -Cert $rootCertPath -FilePath (Join-Path $basePath 'root.crt');                             # public key

		$pfx = New-Object System.Security.Cryptography.X509Certificates.X509Certificate2;
		$pfx.Import((Join-Path $basePath 'root.pfx'), $rootcertPassword, "Exportable,PersistKeySet");
		$store = New-Object System.Security.Cryptography.X509Certificates.X509Store([System.Security.Cryptography.X509Certificates.StoreName]::Root, "localmachine");
		$store.Open("MaxAllowed");
		$store.Add($pfx); 
		$store.Close();
		$store.Dispose();

		# ===========================================================================================================

		Write-Information 'Creating Frontend Server Certificate';

		$frontendCert = New-SelfSignedCertificate -CertStoreLocation cert:\localmachine\my -DnsName @("*.mydomain.com","localhost") -Signer $rootCert -KeyUsage DigitalSignature -KeyUsageProperty All;

		Write-Information 'Exporting Frontend Server Certificate';

		[System.Security.SecureString]$frontendCertPassword = ConvertTo-SecureString -String "PUT SOME REALLY SECURE PASSWORD HERE" -Force -AsPlainText;
		[String]$frontendCertPath = Join-Path -Path 'cert:\localMachine\my\' -ChildPath "$($frontendCert.Thumbprint)";
		Export-PfxCertificate -Cert $frontendCertPath -FilePath (Join-Path $basePath 'FrontendServer/frontend.pfx') -Password $frontendCertPassword; # private key
		Export-Certificate    -Cert $frontendCertPath -FilePath (Join-Path $basePath 'FrontendServer/frontend.crt');                                # public key

		# ===========================================================================================================

		Write-Information 'Creating API Server Certificate';

		$apiCert = New-SelfSignedCertificate -CertStoreLocation cert:\localmachine\my -DnsName @("*.mydomain.com","localhost") -Signer $rootCert -KeyUsage DigitalSignature -KeyUsageProperty All;

		Write-Information 'Exporting API Server Certificate';

		[System.Security.SecureString]$apiCertPassword = ConvertTo-SecureString -String "PUT SOME REALLY SECURE PASSWORD HERE" -Force -AsPlainText;
		[String]$apiCertPath = Join-Path -Path 'cert:\localMachine\my\' -ChildPath "$($apiCert.Thumbprint)";
		Export-PfxCertificate -Cert $apiCertPath -FilePath (Join-Path $basePath 'ApiServer/api.pfx') -Password $apiCertPassword; # private key
		Export-Certificate    -Cert $apiCertPath -FilePath (Join-Path $basePath 'ApiServer/api.crt');                            # public key
    }
}
```

**Comment by jbloem9** [comment]
This issue is now resolved.  Many thanks to all who replied to this issue. I am facing another issue which I have logged #87, around the silent renew giving an error after 10 minutes..

**Comment by RGVGreatCoder** [comment]
Hello. I too gets isAuthorized false every time it returns from successfully authenticating with my IDP (ASPNET Core 1.1/IdentityServer4).  Both, my IDP and Client (ASPNET MVC/Core 2.0/Angular2 from the template which I later upgrade to 4.3.6 to make use of HttpClient) are using HTTPS in my development box, provided by Visual Studio 2017 or IIS Express. Chrome does not seem to complain about such certificate so I suppose I can take it as a "valid-non-public-development-certificate". Why is "isAuthorized" never "true"? Please let me know if I am missing details to better troubleshoot. Thank you and will be waiting for your reply!  ... :(

**Comment by damienbod** [comment]
This usually happens when something is incorrect with the request for Authorization. You can check the F12 console logs in the client browser for the reason. You could also check the logs on hte STS server to see why it was rejected.

Greetings Damien

**Comment by RGVGreatCoder** [comment]
Thank you for your immediate response! 👍 
Well, I don't know where in the debug console log from Visual Studio 2017 says the authentication failed or is incorrect. Could the following pieces of the log (where IDP is InsXIdentity and Client is InsxAngularOidcClient) indicate the authentication went rather successfully?:

    InsXIdentity > info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
    InsXIdentity > Request starting HTTP/ 1.1 GET http://localhost:44333/connect/authorize?client_id=js&redirect_uri=https://localhost:44345/callback&response_type=id_token%20token&scope=openid%20profile%20api1&nonce=N0.54252438313637951505396952853  
    ...
    InsXIdentity > dbug: IdentityServer4.Hosting.EndpointRouter[0]
    InsXIdentity > Mapping found for endpoint: Authorize, creating handler: IdentityServer4.Endpoints.AuthorizeEndpoint
    ...
    InsXIdentity > dbug: IdentityServer4.Endpoints.AuthorizeEndpoint[0]
    InsXIdentity > Start authorize request
    ...
    InsXIdentity > info: IdentityServer4.Endpoints.AuthorizeEndpoint[0]
    InsXIdentity > ValidatedAuthorizeRequest
    ...
    InsXIdentity > [08:49:15 INF] Called login page
    InsXIdentity > info: Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.ViewResultExecutor[1]
    InsXIdentity > Executing ViewResult, running view at path / Views / Account / Login.cshtml.
    InsXIdentity > Request starting HTTP/ 1.1 POST http://localhost:44333/Account/Login?returnurl=%2Fconnect%2Fauthorize%2Flogin%3Fclient_id%3Djs%26redirect_uri%3Dhttps%253A%252F%252Flocalhost%253A44345%252Fcallback%26response_type%3Did_token%2520token%26scope%3Dopenid%2520profile%2520api1%26nonce%3DN0.54252438313637951505396952853 application/x-www-form-urlencoded 251
    InsXIdentity > [08:50:15 INF] In Account Controller
    InsXIdentity > info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
    InsXIdentity > Executing action method InsXIdentity.Controllers.AccountController.Login( InsXIdentity ) with arguments ( InsXIdentity.Models.AccountViewModels.LoginViewModel, /connect/authorize / login ? client_id = js & redirect_uri=https % 3A% 2F% 2Flocalhost% 3A44345% 2Fcallback& response_type=id_token % 20token& scope=openid % 20profile% 20api1& nonce=N0.54252438313637951505396952853) - ModelState is Valid
    InsXIdentity > info: Microsoft.EntityFrameworkCore.Storage.IRelationalCommandBuilderFactory[1]
    InsXIdentity > Executed DbCommand ( 264ms)[Parameters = [@__normalizedUserName_0='?'( Size = 256 )], CommandType = 'Text', CommandTimeout = '30']
    InsXIdentity > SELECT TOP( 1 )[u].[Id], [u].[AccessFailedCount], [u].[ConcurrencyStamp], [u].[Email], [u].[EmailConfirmed], [u].[FirstLoginDate], [u].[LockoutEnabled], [u].[LockoutEnd], [u].[NormalizedEmail], [u].[NormalizedUserName], [u].[PasswordHash], [u].[PhoneNumber], [u].[PhoneNumberConfirmed], [u].[SecurityStamp], [u].[TwoFactorEnabled], [u].[UserName]
    InsXIdentity > FROM[AspNetUsers] AS [u] ...
    ...
    InsXIdentity > [08:50:20 INF] User logged in.
    InsXIdentity > info: Microsoft.AspNetCore.Mvc.Internal.RedirectResultExecutor[1]
    InsXIdentity > Executing RedirectResult, redirecting to / connect / authorize / login ? client_id = js & redirect_uri=https % 3A% 2F% 2Flocalhost% 3A44345% 2Fcallback& response_type=id_token % 20token& scope=openid % 20profile% 20api1& nonce=N0.54252438313637951505396952853.
    ...
    InsXIdentity > info: Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware[8]
    InsXIdentity > AuthenticationScheme: Identity.Application was successfully authenticated.
    InsXIdentity > dbug: IdentityServer4.Hosting.EndpointRouter[0]
    InsXIdentity > Request path / connect / authorize / login matched to endpoint type Authorize
    InsXIdentity> dbug: IdentityServer4.Hosting.EndpointRouter[0]
    InsXIdentity > Mapping found for endpoint: Authorize, creating handler: IdentityServer4.Endpoints.AuthorizeEndpoint
    ...
    InsXIdentity > info: Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware[8]
    InsXIdentity > AuthenticationScheme: Identity.Application was successfully authenticated.
    ...
    InsXIdentity > dbug: IdentityServer4.Validation.AuthorizeRequestValidator[0]
    InsXIdentity > Start authorize request protocol validation
    InsXIdentity > info: Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware[8]
    InsXIdentity > AuthenticationScheme: Identity.Application was successfully authenticated.
    InsXIdentity > dbug: IdentityServer4.Validation.AuthorizeRequestValidator[0]
    InsXIdentity > Calling into custom validator: IdentityServer4.Validation.DefaultCustomAuthorizeRequestValidator
    InsXIdentity > info: IdentityServer4.Endpoints.AuthorizeEndpoint[0]
    InsXIdentity > ValidatedAuthorizeRequest
    InsXIdentity > {
    InsXIdentity>         "ClientId": "js", ...
    ...
    InsXIdentity > dbug: IdentityServer4.ResponseHandling.AuthorizeResponseGenerator[0]
    InsXIdentity>       Creating Implicit Flow response.
    ...
    InsXIdentity > dbug: IdentityServer4.Services.DefaultClaimsService[0]
    InsXIdentity>       In addition to an id_token, an access_token was requested.No claims other than sub are included in the id_token. To obtain more user claims, either use the user info endpoint or set AlwaysIncludeUserClaimsInIdToken on the client configuration.
    InsXIdentity > info: IdentityServer4.Endpoints.AuthorizeEndpoint[0]
    InsXIdentity>       Authorize endpoint response
    InsXIdentity>       {
    InsXIdentity>         "SubjectId": "f30e6c0d-de05-4d35-9d07-448accd60f57",
    InsXIdentity>         "ClientId": "js",
    InsXIdentity>         "RedirectUri": "https://localhost:44345/callback",
    InsXIdentity>         "Scope": "openid profile api1"
    InsXIdentity>       }
    InsXIdentity> info: Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware[8]
    InsXIdentity>       AuthenticationScheme: Identity.Application was successfully authenticated.
    InsXIdentity > info: Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware[8]
    InsXIdentity>       AuthenticationScheme: Identity.Application was successfully authenticated.
    ...
    InsXIdentity > info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
    InsXIdentity>       Request finished in 1782.7665ms 302 
    InsXAngularOidcClient> info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
    InsXAngularOidcClient>       Request starting HTTP/ 1.1 GET http://localhost:44345/callback  
    InsXAngularOidcClient> info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
    InsXAngularOidcClient>       Executing action method InsXAngularOidcClient.Controllers.HomeController.Index( InsXAngularOidcClient ) with arguments (( null ) ) - ModelState is Valid
    InsXAngularOidcClient> info: Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.ViewResultExecutor[1]
    InsXAngularOidcClient>       Executing ViewResult, running view at path / Views / Home / Index.cshtml.
    InsXAngularOidcClient > info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
    InsXAngularOidcClient>       Executed action InsXAngularOidcClient.Controllers.HomeController.Index( InsXAngularOidcClient ) in 1910.0555ms
    InsXAngularOidcClient> info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
    InsXAngularOidcClient>       Request starting HTTP/ 1.1 GET http://localhost:44345/dist/main-client.js?v=yJzp15nrZOD7XwFTtM4TVu2aXockrTPuEZOxF9IJTC0  
    InsXAngularOidcClient> info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
    InsXAngularOidcClient>       Request finished in 2001.6848ms 200 text/html; charset=utf-8
    InsXAngularOidcClient> info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
    InsXAngularOidcClient>       Request finished in 1598.4203ms 200 application/javascript; charset=UTF-8
    InsXIdentity> info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
    InsXIdentity>       Request starting HTTP/ 1.1 GET http://localhost:44333/connect/checksession  
    InsXAngularOidcClient> info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
    InsXAngularOidcClient>       Request starting HTTP/ 1.1 GET http://localhost:44345/dist/__webpack_hmr  
    InsXIdentity> info: Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware[3]
    InsXIdentity>       HttpContext.User merged via AutomaticAuthentication from authenticationScheme: Identity.Application.
    InsXIdentity > info: Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware[8]
    InsXIdentity>       AuthenticationScheme: Identity.Application was successfully authenticated.
    InsXIdentity > dbug: IdentityServer4.Hosting.EndpointRouter[0]
    InsXIdentity>       Request path /connect/checksession matched to endpoint type CheckSession
    InsXIdentity> dbug: IdentityServer4.Hosting.EndpointRouter[0]
    InsXIdentity>       Mapping found for endpoint: CheckSession, creating handler: IdentityServer4.Endpoints.CheckSessionEndpoint
    InsXIdentity> info: IdentityServer4.Hosting.IdentityServerMiddleware[0]
    InsXIdentity>       Invoking IdentityServer endpoint: IdentityServer4.Endpoints.CheckSessionEndpoint for /connect/checksession
    InsXIdentity> dbug: IdentityServer4.Endpoints.CheckSessionEndpoint[0]
    InsXIdentity>       Rendering check session result
    InsXIdentity> info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
    InsXIdentity>       Request finished in 1000.886ms 200 text/html; charset=UTF-8

The IDP (InsXIdentity) has the following client configuration:

    {
        ClientId = "js",
            ClientName = "JavaScript Client",
            AllowedGrantTypes = GrantTypes.Implicit,
            AllowAccessTokensViaBrowser = true,
            RequireConsent = false,

            RedirectUris = new List<string>() { "https://localhost:44345/callback" },
            PostLogoutRedirectUris = new List<string>() { "https://localhost:44345/signout-callback-oidc" },
            AllowedCorsOrigins = new List<string>() { "https://localhost:44345" },

            AllowedScopes =  {
                IdentityServerConstants.StandardScopes.OpenId,
                IdentityServerConstants.StandardScopes.Profile,
                "api1"
            }
    }

The client defines oidcSecureityService as follows:

    constructor( public oidcSecurityService: OidcSecurityService ) {
        const openIDImplicitFlowConfiguration = new OpenIDImplicitFlowConfiguration();
        openIDImplicitFlowConfiguration.stsServer = 'https://localhost:44333/';

        openIDImplicitFlowConfiguration.redirect_url = 'https://localhost:44345/callback';
        openIDImplicitFlowConfiguration.client_id = 'js';
        openIDImplicitFlowConfiguration.response_type = 'id_token token';
        openIDImplicitFlowConfiguration.scope = 'openid profile api1';
        openIDImplicitFlowConfiguration.post_logout_redirect_uri = 'https://localhost:44345/home';
        openIDImplicitFlowConfiguration.start_checksession = true;
        openIDImplicitFlowConfiguration.silent_renew = true;

        openIDImplicitFlowConfiguration.forbidden_route = '/forbidden';
        openIDImplicitFlowConfiguration.unauthorized_route = '/unauthorized';
        openIDImplicitFlowConfiguration.log_console_warning_active = true;
        openIDImplicitFlowConfiguration.log_console_debug_active = false;
        openIDImplicitFlowConfiguration.max_id_token_iat_offset_allowed_in_seconds = 10;

        this.oidcSecurityService.setupModule( openIDImplicitFlowConfiguration );
    }

The browser developer console (F12) did not show any error or failing authorization message. Am I missing something? :( ... Where else can I see if something is incorrect with the request for the Authorization?

**Comment by RGVGreatCoder** [comment]
FIXED!. :) I had commented (yep, smart me) the doCallbaclLogicIfRequired() function because I was having issues with executing "window.location.hash" when trying to run my code for the first time (not knowing what that function did). I removed the comments and yep, it worked. .. :)

I am now facing another issue (when executing this.oidcSecurityService.logoff()) it calls the IDP logout Controller (https://localhost:44333/account/logout) it does not seem to conclude its function. I will check the logs on this one. I suppose I can find this particular issue elsewhere on this site.
